sound/soc/sof/ipc4-pcm.c

```diff
@@ -65,7 +65,7 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 		switch (state) {
 		case SOF_IPC4_PIPE_PAUSED:
 		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
+			if (WIDGET_IS_DAI(swidget->id))
```

Comment:
The original `if (!WIDGET_IS_AIF(swidget->id))` _included_ the new `if (WIDGET_IS_DAI(swidget->id))`, but now the code will also skip AIF for PAUSE/RESET which is against what the comment was saying that the PAUSE/RESET is *only* applied to FE (AIF).

Are you sure that this is correct?

Reply:
> The original `if (!WIDGET_IS_AIF(swidget->id))` _included_ the new `if (WIDGET_IS_DAI(swidget->id))`, but now the code will also skip AIF for PAUSE/RESET which is against what the comment was saying that the PAUSE/RESET is _only_ applied to FE (AIF).
> 
> Are you sure that this is correct?

@ujfalusi When pipeline is set to pause/reset state, the pipeline must contain AIF widget. The patch allow pipeline to change to pause/reset state even it doesn't contain AIF widget. PS: Please see the special case why we need this patch in the comment https://github.com/thesofproject/linux/pull/3866#issuecomment-1251728904.

Reply:
> @libinyang the intention for the !(WIDGET_IS_AIF()) check is to make sure that only the FE pipelines are triggered during the FE CPU DAI trigger. So changing this check is wrong. But I understand where you are coming from. In the case of IPC3, we used the widget's DAPM POWER_UP/DOWN event to trigger the detect pipeline (pipeline 15). So you have 2 choices here: 1. Stick with the same solution as IPC3 2. Expand the check here to also trigger the pipelines associated with widgets that have the same stream name as the host pipeline. I prefer that latter as that will be generic enough to allow other connected pipelines to get triggered as well

@ranj063 Do you mean we should not pause/reset the pipeline 14 in our case in the function sof_ipc4_trigger_pipelines()?  Is there any reason to prevent this? For the second solution you recommended, this requires the widget virtual.detect_sink has the same name as copier.host.13.1. But if there is a case that the widgets like virtual.detect_sink is connected to several widgets, then it's hard to set the stream name of it.

Reply:
Why do we need the same name? Why not set the stream name for all widgets in pipeline 15 to be copier.host.13. And trigger it too because pipeline 13 was triggered

Reply:
@ranj063 I mean if pipeline 15 is also connected to another piepline, let's say pipeline 16 copier.host.16.1, we can't set the pipeline 15 widgets stream name the same as both pipeline 13 and pipeline 16 as ppl13 and ppl16 name may be different.

Reply:
@libinyang my point is we need a generic way to handle intermediate pipelines. host pipelines will be triggered in the FE CPU DAI trigger and DAI pipelines will be triggered in the BE CPU DAI trigger. So the question is what happens to the pipelines in between? And that should be determined by which pipeline they should be trigger alongside with ie host/dai pipeline and this can be indicated with the stream name for the widgets in the intermediate pipelines. So in your example, if there are 2 intermediate pipelines 15 and 16, pipeline 15 would have the stream name copier.host.13 to say that it should be triggered in the FE CPU DAI trigger and pipeline 16 (hypothetically speaking) will have the stream name copier.DMIC.14.1 to say that it should be trigger in the BE CPU DAi trigger

Reply:
@ranj063 I mean the below topology:
ppl 1 --- ppl2 --- ppl3 (host)
                 |-------ppl4 (host)
How can we name the ppl2 widget's stream name, name after ppl3 or ppl4?

And for the solution 1 of using POWER_UP/DOWN event, I still have some concerns:
1. In this case, every intermediate ppl needs such event. It's a little complicated. 
2. intermediate ppl will be triggered start twice. In the POWER_UP event, it will be triggered. And in the sof_ipc4_trigger_pipelines(), it will still be triggered start as we allow all the widgets to be triggered start in this function.
3. We have requirement of the trigger order of the pipelines as Kai mentioned. But if we are using POWER_UP/DOWN event, the trigger order may be wrong.

What do you think if we have a flag in tplg, telling it's a widget on DAI pipline, and we will skip triggering them in the function sof_ipc4_trigger_pipelines()?